File: with-lombok/src/test/java/TestWith.java
import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.With;

class TestWith {

    @Data
    @AllArgsConstructor
    @With
    class User {
        private String firstname;
        private String lastname;
    }

    @Test
    @DisplayName("should clone an instance with With")
    void shouldCloneWithWith() {
        User user1 = new User("John", "Doe");
        User user2 = user1.withLastname("Smith");

        assertThat(user1)
                .hasToString("TestWith.User(firstname=John, lastname=Doe)");

        assertThat(user2)
                .hasToString("TestWith.User(firstname=John, lastname=Smith)");
    }
}

File: with-kotlin/src/test/kotlin/TestWith.kt
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

class TestWith {

    data class User(val firstname: String, val lastname: String)

    @Test
    fun `should clone an instance with copy`() {
        val user1 = User("John","Doe")
        val user2 = user1.copy(lastname = "Smith")

        assertEquals("User(firstname=John, lastname=Doe)",user1.toString())
        assertEquals("User(firstname=John, lastname=Smith)",user2.toString())
    }
}
